data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/InMemoryBuffer.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.accumulator;
+
+import org.opensearch.dataprepper.plugins.s3keyindex.S3ObjectIndex;
+import org.opensearch.dataprepper.plugins.sink.S3SinkConfig;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import java.util.NavigableSet;
+
+/**
+ * Implements the in memory buffer type.
+ */
+public class InMemoryBuffer implements BufferType {
+
+    private S3Client s3Client;
+    private S3SinkConfig s3SinkConfig;
+    private String codecFileExtension;
+
+    public InMemoryBuffer() {
+    }
+
+    /**
+     * @param s3Client
+     * @param s3SinkConfig
+     * @param codecFileExtension
+     */
+    public InMemoryBuffer(final S3Client s3Client, final S3SinkConfig s3SinkConfig, final String codecFileExtension) {
+        this.s3Client = s3Client;
+        this.s3SinkConfig = s3SinkConfig;
+        this.codecFileExtension = codecFileExtension;
+    }
+
+    /**
+     * @param bufferedEventSet
+     * @return boolean
+     * @throws InterruptedException
+     */
+    public boolean inMemoryAccumulate(final NavigableSet<String> bufferedEventSet) throws InterruptedException {
+        String s3ObjectPath = buildingPathPrefix(s3SinkConfig);

```

Comment:
Some of this code is duplicated with the `LocalFileBuffer`. This indicates there should be some refactoring.

Additionally, the decision for building the path is completely independent of the buffering logic. It should probably be moved into its own class which can get the whole object path. Such a small, focused class would be quite easy to unit test as well.

Reply:
Added new class `ObjectKey` for the object path constuction functionality.